[core-http(s)] Expose setClientRequestIdPolicy options

joheredi opened the pull request:
Allow setting `Client Request Id` preferences through pipeline options. This PR enables the user to:

1) Choose to turn off automatically setting `x-ms-client-request-id`
2) Choose a different header name for the ClientRequestId

This is already available to users when sending requests with raw `ServiceClientOptions` instead of `Pipeline`

xirzec commented on sdk/core/core-http/review/core-http.api.md:
```diff
@@ -99,6 +99,12 @@ export class BasicAuthenticationCredentials implements ServiceClientCredentials
+// @public
+export interface ClientRequestIdOptions {
+    requestIdHeaderName?: string;
+    shouldGenerateClientRequestId?: boolean;
```
defaults should be false, so can we make this `disable?: boolean` instead?

xirzec commented on sdk/core/core-http/review/core-http.api.md:
```diff
@@ -525,6 +531,7 @@ export function parseXML(str: string, opts?: SerializerOptions): Promise<any>;
 
 // @public
 export interface PipelineOptions {
+    clientRequestIdOptions?: ClientRequestIdOptions;
```
is this a common consumer scenario? Should we make it an `InternalPipelineOptions` first?

bterlson replied:
This is my main question too - when would a consumer use this? I want to be very sure we want this to be public surface area.

joheredi replied:
Yeah I think this would be better off as an internal property, From the generator standpoint we don't need to have it publicly exposed.

xirzec commented on sdk/core/core-http/src/serviceClient.ts:
```diff
@@ -775,6 +783,12 @@ export function createPipelineFromOptions(
     requestPolicyFactories.push(proxyPolicy(pipelineOptions.proxyOptions));
   }
 
+  if (clientRequestIdOptions.shouldGenerateClientRequestId) {
```
do we want to change the order of this policy? I don't think it'll matter, but I was a little curious why this moved up.

joheredi replied:
Yeah, I'll put it back in the same place it was before

xirzec reviewed (commented):
A few small things, the biggest being on the property that can disable the policy.

Technically in `core-https` someone could do this already by removing the policy by name and adding in a new instance that is configured with the header desired, but I think this is a bit more discoverable.

ramya-rao-a commented:
Is this still applicable?

xirzec commented:
In Corev2 the consumer can just remove the pipeline policy by name, so maybe we don't bother? If this isn't pressing I'd rather not solve it.

ramya-rao-a commented:
@joheredi, Can you share which service package surfaced the need for this change?

deyaaeldeen commented:
@ramya-rao-a this is needed by our code gen testsuite that calls into our test server. @joheredi do you still plan on completing this PR? I can take it up.

joheredi commented:
Yes as @deyaaeldeen mentioned this is for TestServer scenarios in Autorest. This PR was in preparation to the corev2 move, with core-http we can accomplish this with ServiceOptions.

@xirzec, @bterlson, @deyaaeldeen I have rebased master and also addressed the comments, moved these options to InternalPipelineOptions and changed shouldEnable to disable and set the default to false

xirzec commented on sdk/core/core-http/src/serviceClient.ts:
```diff
@@ -802,7 +817,7 @@ export function createPipelineFromOptions(
     keepAlivePolicy(keepAliveOptions),
     userAgentPolicy({ value: userAgentValue }),
-    generateClientRequestIdPolicy(),
+    ...(clientRequestIdPolicy ? [clientRequestIdPolicy] : []),
```
feels weird - why are we putting it into an array just to spread it into a push call?

xirzec commented on sdk/core/core-http/src/policies/generateClientRequestIdPolicy.ts:
```diff
@@ -10,6 +10,29 @@ import {
+}
+
+/**
+ * Default options for ClientRequestId policy
```
this should be `@internal` so it doesn't get picked up by the doc generator right?

xirzec commented on sdk/core/core-rest-pipeline/review/core-rest-pipeline.api.md:
```diff
@@ -272,6 +273,12 @@ export function setClientRequestIdPolicy(requestIdHeaderName?: string): Pipeline
 
+// @public
+export interface SetClientRequestIdPolicyOptions {
+    disable?: boolean;
```
so for `core-rest-pipeline` I did a bunch of work to eliminate `disable` style options, since the pipeline is directly configurable. Can we not simply remove the policy on the pipeline when doing the tests instead of setting this config option?

joheredi replied:
I think this makes sense, we could do this in the generator side
![image](https://user-images.githubusercontent.com/21109913/114480113-b9e3cf80-9bb6-11eb-8b70-6531ee9688b6.png)

xirzec commented on sdk/core/core-rest-pipeline/src/policies/setClientRequestIdPolicy.ts:
```diff
@@ -9,6 +9,29 @@ import { PipelinePolicy } from "../pipeline";
 export const setClientRequestIdPolicyName = "setClientRequestIdPolicy";
 
+/**
+ * Options for how HTTP connections generate the client request id header
```
interface and constant should be `@hidden`

xirzec commented on sdk/core/core-rest-pipeline/src/policies/setClientRequestIdPolicy.ts:
```diff
@@ -9,6 +9,29 @@ import { PipelinePolicy } from "../pipeline";
+/**
+ * Default options for ClientRequestId policy
+ */
+export const DefaultSetClientRequestIdPolicyOptions: SetClientRequestIdPolicyOptions = {
```
feels bad that this options type *isn't* what `setClientRequestIdPolicy` takes as options. This is most of why I think `disable` booleans are bad.

xirzec reviewed (commented):
I'm less fussed about the core-http changes, but I still am not sold on adding a `disable` option to core-rest-pipeline

ramya-rao-a commented:
Side note: Lets hold off on merging this PR until the current set of releases for core packages are out

joheredi commented:
I'll close this PR as this can be accomplished by removing the policy from the pipeline in a cleaner way and we can handle this in the generator side
